### PR TITLE
Replaced hmac.compare_digest with own implementation

### PR DESCRIPTION
**Pull request by lsolodkova**
Answers #16

### CODE REVIEW

File: PA193_mnemonic_Slytherin/mnemonic.py [hunk 1]
```diff
@@ -68,19 +68,13 @@ def __eq__(self, other: object) -> bool:
...
-        # > https://docs.python.org/3.7/library/hmac.html#hmac.compare_digest
-        #
-        # > Note: If a and b are of different lengths, or if an error occurs,
-        # > a timing attack could theoretically reveal information about the types
-        #  > and lengths of a and b—but not their values.
-        #
-        # Type and length of seeds is known to the attacker, but not the value of expected seed.
-        if not isinstance(other, Seed):
+        if not isinstance(other, Seed) or len(self) != len(other):
             return False
```
Comment (mvondracek) [inline PA193_mnemonic_Slytherin/mnemonic.py]: This approach in general would leak information about length of a secret. But in our case, the attacker knows correct length of seed, so it's fine. :)

**Comment by mvondracek** [comment]
@lsolodkova, thanks!

**Review by sobuch — DISMISSED** [review]

File: PA193_mnemonic_Slytherin/mnemonic.py (hunk 1 above)
Comment (lsolodkova) [reply PA193_mnemonic_Slytherin/mnemonic.py]: I've replaced quick return, please, check one more time.

File: PA193_mnemonic_Slytherin/mnemonic.py [hunk 2]
```diff
@@ -68,12 +68,14 @@ def __eq__(self, other: object) -> bool:
         :rtype: bool
         :return: True if seeds are the same, False otherwise.
         """
+        result = 0
         if not isinstance(other, Seed) or len(self) != len(other):
-            return False
```
Comment (sobuch) [inline PA193_mnemonic_Slytherin/mnemonic.py:82]: is it safe to continue if type(other)!=Seed

File: PA193_mnemonic_Slytherin/mnemonic.py (hunk 1 above)
Comment (mvondracek) [reply PA193_mnemonic_Slytherin/mnemonic.py]: If it's valid instance of `Seed` then the length is fixed `SEED_LEN`.

File: PA193_mnemonic_Slytherin/mnemonic.py [hunk 3]
```diff
@@ -68,19 +68,13 @@ def __eq__(self, other: object) -> bool:
...
-        #  > it appropriate for cryptography
-        # > https://docs.python.org/3.7/library/hmac.html#hmac.compare_digest
-        #
-        # > Note: If a and b are of different lengths, or if an error occurs,
-        # > a timing attack could theoretically reveal information about the types
-        #  > and lengths of a and b—but not their values.
-        #
-        # Type and length of seeds is known to the attacker, but not the value of expected seed.
-        if not isinstance(other, Seed):
+        if not isinstance(other, Seed) or len(self) != len(other):
```
Comment (mvondracek) [inline PA193_mnemonic_Slytherin/mnemonic.py]: If it's valid instance of `Seed` then the length is fixed `SEED_LEN`.

File: PA193_mnemonic_Slytherin/mnemonic.py [hunk 4]
```diff
@@ -68,19 +68,15 @@ def __eq__(self, other: object) -> bool:
...
-        return hmac.compare_digest(self, other)
+        result = 0
+        if not isinstance(other, Seed) or len(self) != len(other):
+            result = 1
+            s = self
+        else:
+            s = other
+        for b1, b2 in zip(self, s):
+            result |= b1 ^ b2
+        return result == 0
```
Comment (mvondracek) [inline PA193_mnemonic_Slytherin/mnemonic.py:79]: I think this is good.

File: PA193_mnemonic_Slytherin/mnemonic.py (hunk 1 above)
Comment (lsolodkova) [reply PA193_mnemonic_Slytherin/mnemonic.py]: Removed extra length check, thanks!

**Review by mvondracek — APPROVED** [review]

**Review by sobuch — APPROVED** [review]

**Pull request merged**